Add the Times system to keep track of time dynamically

GiudGiud opened the pull request:
Some starter objects to get going with the capability
and some APIs as well that I think make sense.

Modeled after Positions

An alternative design would have been to template on the data type, something like "DataBase" base class that we instantiate with Real / Point / whatever is needed
While it is satisfying to have inheritance trees, the shared code amount is not large here. There's a lot of differences in the APIs. "nearest" is not a big deal like for Positions. "previous/next" is better defined, and there's no 3D/4D structure imaginable

maybe next time the need arises

GiudGiud committed:
Add a base class for Times.
Similar to Positions, but uses Real

GiudGiud committed:
Add an object to get times from the input file

GiudGiud committed:
Add times from CSV

GiudGiud committed:
Add times from a reporter / another times object

GiudGiud committed:
Add times from an exodus file

GiudGiud committed:
misc doco fixup

GiudGiud committed:
Add a times object where we evaluate a functor
refs #24696

GiudGiud committed:
Add a Times object from the times taken by the executioner

GiudGiud committed:
Add capability to take Times objects as parameters

GiudGiud committed:
Add ability to use Times as sync times for output

GiudGiud committed:
Add unit test for Times base class routines, refs #24696

GiudGiud committed:
Add a syntax page for the Times system
Use it for the link in source index
refs #24696

GiudGiud committed:
Add documentation to all times objects, refs #24696

GiudGiud committed:
Add a time stepper that relies on times
refs #24696

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1614565/) on 8286239 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24818/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24818/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0e3d4f264182cdd0d2b4392368092302fad62522`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1614565/) on 8286239 wanted to post the following:

The following commit(s) contain fixups:

- 8c33d5a73f: fixup! Add a times object where we evaluate a functor refs #24696

GiudGiud committed:
Add testing for all the times objects implemented
refs #24696

GiudGiud committed:
Add action to add the times objects

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1614615/) on f2f0bda wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24818/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24818/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0e3d4f264182cdd0d2b4392368092302fad62522`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1614659/) on e50e759 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24818/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24818/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0e3d4f264182cdd0d2b4392368092302fad62522`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1614714/) on 7c394d3 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24818/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24818/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0e3d4f264182cdd0d2b4392368092302fad62522`

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1614932/) on 17811fc wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>0e3d4f</th>
      <th colspan=4>#24818 17811f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.58%</td>
      <td>85.59%</td>
      <td>+0.00%</td>
      <td>86.67%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>89021</td>
      <td>89210</td>
      <td>+189</td>
      <td>221</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14997</td>
      <td>15024</td>
      <td>+27</td>
      <td>34</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24818/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24818/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24818/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24818/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24818/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24818/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24818/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24818/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24818/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24818/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24818/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24818/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24818/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24818/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24818/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24818/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24818/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24818/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24818/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24818/coverage/porous_flow)
- `ray_tracing` (FAILED)
- `rdg` (FAILED)
- `reactor` (FAILED)
- `richards` (FAILED)
- `scalar_transport` (FAILED)
- `solid_properties` (FAILED)
- `stochastic_tools` (FAILED)
- `tensor_mechanics` (FAILED)
- `thermal_hydraulics` (FAILED)
- `xfem` (FAILED)

</details>

## Warnings

- `framework` new line coverage rate 86.67% is less than the suggested 90.0%



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1614932/) on 17811fc wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `rdg` coverage rate (required: 64.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `richards` coverage rate (required: 94.0%)
- Failed to generate `scalar_transport` coverage rate (required: 90.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `tensor_mechanics` coverage rate (required: 84.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 90.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1614891/) on 17811fc wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24818/site)

This comment will be updated on new commits.

GiudGiud commented on framework/doc/content/source/functions/PiecewiseLinearFromVectorPostprocessor.md:
```diff
@@ -0,0 +1,21 @@
+# PiecewiseLinearFromVectorPostprocessor
```
total rider commit. please let me know if it s not welcome

lindsayad replied:
I'd appreciate separation

GiudGiud replied:
sounds good, I ll move the commit to a general doco PR

lindsayad commented on framework/doc/content/source/times/ExodusFileTimes.md:
```diff
@@ -0,0 +1,13 @@
+
+!syntax description /Times/ExodusFileTimes
+
+The exodus files will only be read on the first parallel process, and the times
```
```suggestion
The exodus files will only be read on the zeroth parallel process, and the times
```

lindsayad commented on framework/doc/content/source/times/ExodusFileTimes.md:
```diff
@@ -0,0 +1,13 @@
+!syntax description /Times/ExodusFileTimes
+
+The exodus files will only be read on the first parallel process, and the times
+will be broadcast to all ranks.
```
```suggestion
will be broadcast to all other ranks.
```

lindsayad commented on framework/doc/content/source/times/FunctorTimes.md:
```diff
@@ -0,0 +1,15 @@
+
+!syntax description /Times/FunctorTimes
+
+The functor and its factor will be evaluated in the element containing the (0,0,0) point, the latter of which should lay
```
```suggestion
The functor and its factor will be evaluated in the element containing the (0,0,0) point, the latter of which should lie
```

lindsayad commented on framework/doc/content/source/times/FunctorTimes.md:
```diff
@@ -0,0 +1,15 @@
+
+!syntax description /Times/FunctorTimes
+
+The functor and its factor will be evaluated in the element containing the (0,0,0) point, the latter of which should lay
```
This seems bizarre, but maybe I'll see the logic of it later

GiudGiud replied:
well we need an element to evaluate a functor.

maybe there should be a "default argument" that does not involve anything (no face, elem, side, node) to functors?

lindsayad replied:
An "Anywhere" spatial argument that only some derived functors can handle does sound like a logical idea to me

lindsayad commented on framework/doc/content/source/timesteppers/TimeSequenceFromTimes.md:
very neat

lindsayad commented on framework/include/times/CSVFileTimes.h:
```diff
@@ -0,0 +1,30 @@
+  CSVFileTimes(const InputParameters & parameters);
+  virtual ~CSVFileTimes() = default;
+
+  void initialize() override {}
```
```suggestion
  virtual void initialize() override {}
```

lindsayad commented on framework/include/times/ExodusFileTimes.h:
```diff
@@ -0,0 +1,26 @@
+  ExodusFileTimes(const InputParameters & parameters);
+  virtual ~ExodusFileTimes() = default;
+
+  void initialize() override {}
```
```suggestion
  virtual void initialize() override {}
```

lindsayad commented on framework/include/times/FunctorTimes.h:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Times created by evaluating a functor
+ */
+class FunctorTimes : public Times, public NonADFunctorInterface
```
nice use of the `NonADFunctorInterface`!

lindsayad commented on framework/include/times/FunctorTimes.h:
```diff
@@ -0,0 +1,33 @@
+  FunctorTimes(const InputParameters & parameters);
+  virtual ~FunctorTimes() = default;
+
+  void initialize() override;
```
```suggestion
  virtual void initialize() override;
```

lindsayad commented on framework/include/times/InputTimes.h:
```diff
@@ -0,0 +1,26 @@
+  InputTimes(const InputParameters & parameters);
+  virtual ~InputTimes() = default;
+
+  void initialize() override {}
```
```suggestion
  virtual void initialize() override {}
```

lindsayad commented on framework/include/times/ReporterTimes.h:
```diff
@@ -0,0 +1,26 @@
+  ReporterTimes(const InputParameters & parameters);
+  virtual ~ReporterTimes() = default;
+
+  void initialize() override;
```
```suggestion
  virtual void initialize() override;
```

lindsayad commented on framework/include/times/SimulationTimes.h:
```diff
@@ -0,0 +1,26 @@
+  SimulationTimes(const InputParameters & parameters);
+  virtual ~SimulationTimes() = default;
+
+  void initialize() override;
```
```suggestion
  virtual void initialize() override;
```

lindsayad commented on framework/include/times/Times.h:
```diff
@@ -0,0 +1,76 @@
+  Times(const InputParameters & parameters);
+  virtual ~Times() = default;
+
+  /// Getter for the full times vector
+  const std::vector<Real> & getTimes() const;
+
+  /// Getter for a set of the full times
+  const std::set<Real> getUniqueTimes() const
+  {
+    return std::set<Real>(_times.begin(), _times.end());
+  };
```
what is the value in having duplicate entries? I guess my inclination would be that we just have the set

GiudGiud replied:
keep our options open?
and sets are both harder to work with as a dev, and slower in the code

lindsayad replied:
What APIs on a set are confusing? And speed really depends on what you're doing. If you have to sort a vector a lot, then that will be slow. If you need to search for something in a vector vs a set, its O(N) vs. O(log N)

GiudGiud replied:
you cant index a set directly for example

my experience with sets is that you always end up converting them to vectors somewhere along the way

"uniqueness" in floating point numbers is also not very well defined with the floating point comparisons issues

lindsayad replied:
If you have multiple sources though, or even a single source, what is the usefulness of a `getTimeAtIndex` API? Seems like the `getPreviousTime` and `getNextTime` are the useful APIs and those are only useful if your data is sorted

lindsayad replied:
> my experience with sets is that you always end up converting them to vectors somewhere along the way

I don't share this experience

lindsayad replied:
And those APIs would require container searches which is O(N) for vectors

GiudGiud replied:
> > my experience with sets is that you always end up converting them to vectors somewhere along the way
> 
> I don't share this experience

Checkout the block restrictable interface. Under the hood it's a set, and there's vector conversions APIs all over.

lindsayad replied:
Sure that's an instance of it. But that's because people designed a lot of APIs that expected vectors instead of sets. Poor API design. In the end we store both the moose and libmesh subdomains as sets

roystgnr replied:
"Build a `set` then convert to a sorted `vector`" is almost a C++ design pattern, as the fastest way to handle the "do lots of inserts with duplication, then do lots of searches" use case.

But if neither the inserts nor the searches are in an inner kernel then the speed of sorted vectors doesn't matter.

I'm also not seeing what the use is of getting a time by index.  Knowing an index would imply knowing about every other bit of code that might be adding times, right?  And half the point of this system is to make that unnecessary?

lindsayad commented on framework/include/times/Times.h:
```diff
@@ -0,0 +1,76 @@
+protected:
+  /// In charge of computing / loading the times, unless all that could be done there is done
+  /// in the constructor
+  virtual void initialize() override = 0;
```
If it's protected in the base, let's make this protected in the derived classes as well

GiudGiud replied:
I think they are public in the base?

lindsayad replied:
This is `protected`. It says so right here. Derived classes of `Times` are overriding it in the public access space though

GiudGiud replied:
fixed in derived classes

lindsayad commented on framework/include/times/Times.h:
```diff
@@ -0,0 +1,76 @@
+
+  /// Whether generation of times is distributed or not (and therefore needs a broadcast)
+  bool _need_broadcast;
+  /// Whether times should be sorted, because they come from different sources for example
+  bool _need_sort;
+  /// Whether duplicate times should be removed
+  bool _need_unique;
```
would be unnecessary if we just use a set

roystgnr replied:
Even with `std::set`, I think @GiudGiud is onto something with the "floating point comparisons issues" comment in another thread ... but if we're expecting to be computing times with FP roundoff error, then do we want "uniqueness" up to some specified tolerance, not just a bool?  A `set` will give unique times with exact comparisons, but if we have a `t=0.2` and another `t=0.02*10` then we'll need some `epsilon > 3e-17` to tell the system that we really think those two unique-in-FP numbers should be considered equivalent.

lindsayad replied:
It's not really about uniqueness to me though, it's about whether it's sorted

lindsayad replied:
But the concern you raise is valid

GiudGiud replied:
I ll put up some patch for tolerance. 

I dont think I want them sorted by default.

lindsayad replied:
why not?

GiudGiud replied:
see reasons above plus it's really easy to think of a use case where we would want more than one time. Let me give examples here and I can come up with more if still not convinced 

- if we were to keep track of when a simulation failed. The first time it failed a step could be at t=100, we try again with a cut step at t=50. If we fail again then pass at 25, we loose track of how it unfolded. If we kept times unique, we can even loose track of how many times we failed a certain time

- we have multiple subapps contributing times to a Times object that is keeping track of an event. Subapp 1 says t=25, subapp 2 says t=20. Then if we sort, we loose track that subapp 2 had the lower number

- we run a transient direct then an adjoint calculation (with the reverse time). If we sort, we loose track that the times in the adjoint simulation came after, they will all be mixed together

lindsayad replied:
alright, those are great examples 😄

lindsayad replied:
A perfect use case for a `std::vector`: a sequence!

lindsayad commented on framework/include/timesteppers/TimeSequenceFromTimes.h:
```diff
@@ -0,0 +1,31 @@
+class Times;
+
+/**
+ * Time sequence stepper that gets it sequence of times to hit from a times object
```
```suggestion
 * Time sequence stepper that gets its sequence of times to hit from a times object
```

lindsayad reviewed (changes requested):
Went through doc and headers

GiudGiud committed:
Apply suggestions from code review
- docs
- explcitly call virtual before override
- keep consistent level of publicness in inheriting classes

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

GiudGiud committed:
Add an absolute tolerance on times to handle uniqueness concerns,  refs #24696

lindsayad commented on framework/include/times/Times.h:
```diff
@@ -0,0 +1,78 @@
+  const std::vector<Real> & getTimes() const;
+
+  /// Getter for a set of the full times
+  const std::set<Real> getUniqueTimes() const
```
It's not usual practice to return a `const` value, but it would be unusual for a user to want to modify this return value, so it's probably fine

lindsayad commented on framework/src/outputs/Output.C:
```diff
@@ -117,6 +122,10 @@ Output::Output(const InputParameters & parameters)
     _sync_times(std::set<Real>(getParam<std::vector<Real>>("sync_times").begin(),
                                getParam<std::vector<Real>>("sync_times").end())),
+    _sync_times_object(isParamValid("sync_times_object")
+                           ? dynamic_cast<Times *>(&_problem_ptr->getUserObject<Times>(
```
```suggestion
                           ? static_cast<Times *>(&_problem_ptr->getUserObject<Times>(
```
If the casting fails it will fail first in `getUserObject` to it should be safe to `static_cast` here

lindsayad commented on framework/src/outputs/Output.C:
```diff
@@ -174,6 +183,14 @@ Output::Output(const InputParameters & parameters)
   }
+
+  // Check that the sync times were retrieved as expected
+  if (isParamSetByUser("sync_times_object") && !_sync_times_object)
+    paramError("sync_times_object", "Times object does not exist");
```
should be able to remove this as if casting fails it will fail in the `getUserObject` call. But if you want to be doubly safe and choose to keep this, then ignore my `static_cast` suggestion above

GiudGiud replied:
```suggestion
```

lindsayad commented on framework/src/times/CSVFileTimes.C:
```diff
@@ -0,0 +1,46 @@
+    file.setFormatFlag(MooseUtils::DelimitedFileReader::FormatFlag::COLUMNS);
+    file.read();
+
+    const std::vector<std::vector<double>> & data = file.getData();
```
```suggestion
    const auto & data = file.getData();
```

lindsayad commented on framework/src/times/ExodusFileTimes.C:
```diff
@@ -0,0 +1,55 @@
+  InputParameters params = Times::validParams();
+  params.addClassDescription("Import times from one or more Exodus files.");
+  params.addRequiredParam<std::vector<FileName>>("files", "Exodus file(s) with the times");
+  // File is loaded only on 1st process
```
```suggestion
  // File is loaded only on zeroth process
```

lindsayad commented on framework/src/times/FunctorTimes.C:
```diff
@@ -0,0 +1,57 @@
+void
+FunctorTimes::initialize()
+{
+  // Locate the origin on the mesh
+  Point p(0, 0, 0);
+  auto pl = _fe_problem.mesh().getMesh().sub_point_locator();
+  auto * elem = (*pl)(p);
+  if (!elem)
+    mooseError("Origin point not in local mesh, cannot evaluate the functor there");
```
What if you just chose the ID 0 element? @roystgnr what happens with point locators on distributed meshes? My suggestion will only be globally consistent on replicated meshes. Wondering whether this current implementation has the same issue

roystgnr replied:
If the point is on a local or ghosted element, the point locator will return one of those.  If it's only on remote elements, the point locator will return null.

An element with id 0 might not exist either; element deletion and `allow_renumbering(false)` are things.

If you need an arbitrary element on every processor that'll be computing on any elements then `*(mesh.elements_begin())` will get you one (though you'll still need to check that `elements_begin() != elements_end()` to handle distributed meshes where a processor has no elements).

If you need an arbitrary element on every processor and it's the same element on every processor ... this can't be done in general.

roystgnr replied:
Would it make sense to only evaluate on processor(s) which contain `Point(0)`, then find the lowest-ranked one of those and `broadcast()` from there?

GiudGiud replied:
Second solution makes sense. 
However the real solution here is a MeshArg argument to functor. 
For all the functors for which it would be defined (functions? with an extra parameter maybe to specify the point)

Do we really need this now though

lindsayad replied:
> Do we really need this now though

I don't know to what you're referring to, but I'm fine leaving this as is as long as you add an error check or at least a warning check in this class for whether the mesh is distributed

GiudGiud replied:
added an errorIfDistributed

lindsayad commented on framework/src/times/FunctorTimes.C:
```diff
@@ -0,0 +1,57 @@
+    mooseError("Origin point not in local mesh, cannot evaluate the functor there");
+  Moose::ElemArg elem_origin = makeElemArg(elem);
+
+  auto t = determineState();
```
```suggestion
  const auto t = determineState();
```

lindsayad commented on framework/src/times/ReporterTimes.C:
```diff
@@ -0,0 +1,58 @@
+
+  for (const auto r_it : index_range(positions_reporters))
+  {
+    const std::string & reporter_name = positions_reporters[r_it];
```
```suggestion
    const auto & reporter_name = positions_reporters[r_it];
```

lindsayad commented on framework/src/times/SimulationTimes.C:
```diff
@@ -0,0 +1,39 @@
+  params.set<bool>("auto_broadcast") = false;
+  // Times history of a simulation is already one-way
+  params.set<bool>("auto_sort") = false;
+  // Timestep_begin is as early and as often as we need
```
```suggestion
  // TIMESTEP_BEGIN is as early and as often as we need
```

lindsayad commented on framework/src/times/SimulationTimes.C:
```diff
@@ -0,0 +1,39 @@
+void
+SimulationTimes::initialize()
+{
+  // Initialize is by default what is called by ::execute()
```
```suggestion
  // Initialize is by default what is called by execute()
```

lindsayad commented on framework/src/times/SimulationTimes.C:
```diff
@@ -0,0 +1,39 @@
+  // Initialize is by default what is called by ::execute()
+  _times.push_back(_fe_problem.time());
+  // if this is performed multiple times (fixed point iterations)
+  // it will be caught by our logic to make the vector hold unique times
```
In the comment can you point to the function(s) that handle this?

GiudGiud replied:
```suggestion
  // it will be caught by our logic to make the vector hold unique times in finalize()
```

lindsayad commented on framework/src/times/Times.C:
```diff
@@ -0,0 +1,124 @@
+  params.addParam<Real>(
+      "unique_tolerance",
+      1e-12,
+      "Absolute tolerance for removing duplicated times when using setting unique_times to true");
```
```suggestion
      "Absolute tolerance for removing duplicated times when getting a set of unique times");
```

lindsayad commented on framework/src/times/Times.C:
```diff
@@ -0,0 +1,124 @@
+    mooseError("Times vector has not been initialized.");
+
+  Real previous_time = _times[0];
+  for (const auto i : make_range(_times.size() - 1))
+  {
+    const auto & time = _times[i + 1];
```
```suggestion
  for (const auto i : make_range(std::size_t(1), _times.size()))
  {
    const auto & time = _times[i];
```

lindsayad commented on framework/src/times/Times.C:
```diff
@@ -0,0 +1,124 @@
+}
+
+Real
+Times::getPreviousTime(const Real current_time) const
```
Given that `_times` is a sequence, there could in theory be multiple "previous" times for a given time, right? I would put this somewhere in the documentation

lindsayad commented on framework/src/times/Times.C:
```diff
@@ -0,0 +1,124 @@
+}
+
+Real
+Times::getNextTime(const Real current_time) const
```
Same comment as for `getPreviousTime` regarding multiple possible "next times"

lindsayad commented on framework/src/timesteppers/TimeSequenceFromTimes.C:
```diff
@@ -0,0 +1,43 @@
+    _times(_fe_problem.getUserObject<Times>(getParam<TimesName>("times")))
+{
+  // If they are available, initialize
+  auto times = _times.getTimes();
```
```suggestion
  const auto & times = _times.getTimes();
```

lindsayad commented on framework/src/timesteppers/TimeSequenceFromTimes.C:
```diff
@@ -0,0 +1,43 @@
+TimeSequenceFromTimes::step()
+{
+  // Get the times again in case there are new ones
+  auto times = _times.getTimes();
```
```suggestion
  const auto & times = _times.getTimes();
```

lindsayad commented on test/tests/times/tests:
```diff
@@ -0,0 +1,47 @@
+      input = 'functor_times.i'
+      jsondiff = 'functor_times_out.json'
+      detail = 'a functor evaluated at the element containing the origin,'
+      # Needs to be able to find the element containing the origin
```
Ah this answered my question from before ... point locator will only be globally consistent on replicated mesh

lindsayad commented on test/tests/times/tests:
```diff
@@ -0,0 +1,47 @@
+      type = 'JSONDiff'
+      input = 'reporter_times.i'
+      jsondiff = 'reporter_times_out.json'
+      detail = 'a reporter with data in the expected floating point vector format and'
```
```suggestion
      detail = 'a reporter with data in the expected floating point vector format, and'
```

GiudGiud commented on framework/include/times/Times.h:
```diff
@@ -0,0 +1,78 @@
+
+  /// Find the previous time in the times vector for a given time
+  /// If current_time is also in the times vector within numerical precision, will return the previous value
+  Real getPreviousTime(const Real current_time) const;
```
```suggestion
  /// If the times are not sorted or unique, this will return the time right before the first time found above the current_time
  Real getPreviousTime(const Real current_time) const;
```

GiudGiud commented on framework/include/times/Times.h:
```diff
@@ -0,0 +1,78 @@
+
+  /// Find the next time in the times vector for a given time
+  /// If current_time is also in the times vector within numerical precision, will return the next value
+  Real getNextTime(const Real current_time) const;
```
```suggestion
  /// If the times are not sorted or unique, this will return the first time above the current_time in the
  /// times vector
  Real getNextTime(const Real current_time) const;
```

GiudGiud committed:
Address Alex review

- warn in APIs about potential issues with non-uniquess and non-sorting
- use static_cast from the getUO since we know it will be cast well
- dont allow distributed meshes from the get-go with Functor Times

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

GiudGiud commented:
Thanks for the review!
